Second review of mail test module...

the-t-in-rtf opened the pull request:
I addressed the feedback from the first review, here is a sane pull request from my branch to GPII's master.

the-t-in-rtf committed:
Initial commit for second pull request.

the-t-in-rtf committed:
Saved changes to README

the-t-in-rtf committed:
Cleaned up after merge

the-t-in-rtf committed:
Moved nodemailer dependencies into main block.

the-t-in-rtf committed:
Fixed package name to match new GPII repo.

the-t-in-rtf committed:
Removed outdated (and misfiled) mail handler based on old conventions.

the-t-in-rtf committed:
Ignored node_modules directory.

the-t-in-rtf committed:
Simplified harness setup to address problems discovered while working with express-couchuser module.

the-t-in-rtf committed:
Fixed wiring for messageFile so that the defaults are acceptable for most tests.

the-t-in-rtf committed:
Continued working on messageFile variable visibility.  Apparently "members" is a poor place for it to live.

the-t-in-rtf committed:
Added Gruntfile.js to enable all common grunt tasks to be run against this project.

the-t-in-rtf committed:
Fixed all errors reported by JSHint.

the-t-in-rtf committed:
Created `src/js/index.js` to make it easier to require all dependencies.

the-t-in-rtf committed:
Updated tests to fix all JSHint errors.

the-t-in-rtf committed:
Moved index.js to allow "requiring" of entire package.

the-t-in-rtf committed:
Started addressing feedback from pull request.

the-t-in-rtf committed:
Updated tests per feedback from Antranig.

the-t-in-rtf committed:
Cleaned up unneeded test options following latest refactor.

amb26 commented on src/js/mailserver.js:
```diff
@@ -0,0 +1,55 @@
+            "SMTPBanner":           "Test Mail Server",
+            "queueID":              "TESTMAIL",
+            "disableDNSValidation": true,
+            "outputDir":            "/tmp",
```
This path is not valid on Windows. node.js exposes an API for this purpose - https://nodejs.org/api/os.html#os_os_tmpdir

the-t-in-rtf replied:
This has been addressed as suggested.

amb26 commented on src/js/simpleSmtpServer.js:
```diff
@@ -0,0 +1,84 @@
+
+gpii.test.mail.smtp.simpleSmtpServer.handleStartData = function (that, connection) {
+    var timestamp = (new Date()).getTime();
+    that.options.messageFile = that.options.config.outputDir + "/message-" + timestamp + ".txt";
```
Writing directly into a component's options from code will be prohibited. Either write an "expander" for this purpose or else you can just write it into the component's top level

the-t-in-rtf replied:
Someone else mentioned "members" for this type of thing, but you've continued to push for top-levelling things like this.  I'd like to know the history of the "members" pattern if you know it.

amb26 replied:
The effect of "members" is what I mean by this - options registered in "members" become top-level component material - that is, the effect is the same as writing manual code which assigns to "that.x". Top-level material in "options" goes into "that.options" in the final component and should be considered constant.

the-t-in-rtf replied:
Thanks for the explanation, it all makes sense now.

the-t-in-rtf replied:
Fixed.

amb26 commented on src/js/simpleSmtpServer.js:
```diff
@@ -0,0 +1,84 @@
+
+    that.events.messageReceived.fire(that, connection);
+
+    callback(null, that.options.config.queueId);
```
Instead of being applied in the same callback, this should be registered in a separate listener with priority: "last" to ensure that any other listeners that users register to this event are notified before we call the mysterious "callback" supplied from the native dataReady event. At least, this is the purpose I understand for this callback - I assume it is intended as a notification to the implementation that we have finished all our processing for the data it supplied.

the-t-in-rtf replied:
Yes, in looking at this I can see that the callback is a bad pattern.  I'm also considering having a separate named event that is meant to be bound to instead of the raw events that mirror simplesmtp's lifecycle.

the-t-in-rtf replied:
In looking at this further, I am confused about how to break this down.  I have to register my function with simplesmtp.on('event_name', function).  I don't see how I would convert that to working more cleanly as a series of listeners.

amb26 replied:
I don't see any code resembling that in the function I am commenting - could you be more specific?

amb26 replied:
Sorry, yes, I see the problem now. This callback pattern does make sequencing problematic.

amb26 replied:
The code is best left the way it is, in that case - I can't think of a better solution

the-t-in-rtf replied:
We ended up agreeing not to do this and I reverted my interim approach.

amb26 commented on src/js/simpleSmtpServer.js:
```diff
@@ -0,0 +1,84 @@
+            "funcName": "gpii.test.mail.smtp.simpleSmtpServer.handleData",
+            "args": ["{that}", "{arguments}.0", "{arguments}.1"]
+        },
+        "handleDataReady": {
```
Namespaced listener is probably more helpful than invoker - especially since we will now need two of these - "handleDataReady.messageReceived" and "handleDataReady.complete"

the-t-in-rtf replied:
This is no longer an option, we need named invokers so that simplesmtp.on calls will include {that}.

the-t-in-rtf committed:
Made changes based on pull request feedback.

the-t-in-rtf commented:
@amb26, I have addressed all of your feedback.  The only remaining problem I am aware of is that the tests are not firing correctly (only one of two tests are even executed).

I will plan to refactor that next week, if you could give all-tests.js a quick review, I would appreciate it.

the-t-in-rtf committed:
Sanitized used of top-level variables so that `members` are now used consistently.

the-t-in-rtf commented:
The tests are now firing and all feedback has been responded to.

amb26 commented on src/js/simpleSmtpServer.js:
```diff
@@ -0,0 +1,86 @@
+
+gpii.test.mail.smtp.simpleSmtpServer.handleStartData = function (that, connection) {
+    var timestamp = (new Date()).getTime();
+    that.messageFile = that.options.config.outputDir + "/message-" + timestamp + ".txt";
```
Should rename this something like "currentMessageFile" to emphasize that it is some volatile piece of state that will keep getting overwritten with each new connection.

the-t-in-rtf replied:
Done.

duhrer committed:
Renamed `messageFile` per feedback from pull request.

duhrer committed:
Added "ready" message to assist in debugging lifecycle issues elsewhere.  Can be disabled or overridden by replacing the named listener.

duhrer committed:
Converted all uses of `console.log` to use `fluid.log` instead.
